lib/smf/smf.c

```diff
@@ -199,98 +239,150 @@ void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state)
+	}
+#else
+	/* execute entry action if it exists */
 	if (init_state->entry) {
 		init_state->entry(ctx);
 	}
+#endif
 }
 
-void smf_set_state(struct smf_ctx *const ctx, const struct smf_state *target)
+void smf_set_state(struct smf_ctx *const ctx, const struct smf_state *new_state)
 {
-	struct internal_ctx * const internal = (void *) &ctx->internal;
+	struct internal_ctx *const internal = (void *)&ctx->internal;
+
+	if (new_state == NULL) {
+		LOG_ERR("new_state cannot be NULL");
+		return;
+	}
 
 	/*
-	 * It does not make sense to call set_state in an exit phase of a state
+	 * It does not make sense to call smf_set_state in an exit phase of a state
 	 * since we are already in a transition; we would always ignore the
 	 * intended state to transition into.
 	 */
-	if (internal->exit) {
-		LOG_WRN("Calling %s from exit action", __func__);
+	if (internal->is_exit) {
+		LOG_ERR("Calling %s from exit action", __func__);
 		return;
 	}
 
-	internal->exit = true;
+#ifdef CONFIG_SMF_ANCESTOR_SUPPORT
+	const struct smf_state *topmost;
+
+	if (share_paren(ctx->executing, new_state)) {
+		/* new state is a parent of where we are now*/
+		topmost = new_state;
+	} else if (share_paren(new_state, ctx->executing)) {
+		/* we are a parent of the new state */
+		topmost = ctx->executing;
+	} else {
+		/* not directly related, find LCA */
+		topmost = get_lca_of(ctx->executing, new_state);
+	}
 
-	/* Execute the current states exit action */
-	if (ctx->current->exit) {
-		ctx->current->exit(ctx);
+	internal->is_exit = true;
+	internal->new_state = true;
 
-		/*
-		 * No need to continue if terminate was set in the
-		 * exit action
-		 */
+	/* call all exit actions up to (but not including) the topmost */
+	if (smf_execute_all_exit_actions(ctx, topmost)) {
+		/* No need to continue if terminate was set in the exit action */
+		return;
+	}
+
+	/* if self-transition, call the exit action */
+	if ((ctx->executing == new_state) && (new_state->exit)) {
+		new_state->exit(ctx);
+
+		/* No need to continue if terminate was set in the exit action */
 		if (internal->terminate) {
 			return;
 		}
 	}
 
-	if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT)) {
-		internal->new_state = true;
+	internal->is_exit = false;
 
-		if (smf_execute_ancestor_exit_actions(ctx, target)) {
+	/* if self transition, call the entry action */
+	if ((ctx->executing == new_state) && (new_state->entry)) {
+		new_state->entry(ctx);
+
+		/* No need to continue if terminate was set in the entry action */
+		if (internal->terminate) {
 			return;
 		}
 	}
-
-	internal->exit = false;
-
 #ifdef CONFIG_SMF_INITIAL_TRANSITION
-	/*
-	 * The final target will be the deepest leaf state that
-	 * the target contains. Set that as the real target.
-	 */
-	while (target->initial) {
-		target = target->initial;
-	}
+		/*
+		 * The final target will be the deepest leaf state that
+		 * the target contains. Set that as the real target.
+		 */
+		while (new_state->initial) {
+			new_state = new_state->initial;
+		}
```

Comment:
Fix the indentation.

Reply:
Drat. Thought i caught them all. Clang-format run on the file.